exporter/elasticsearchexporter/internal/objmodel/objmodel.go

```diff
@@ -0,0 +1,412 @@
+// map. All nested maps will be flattened, with keys being joined using a `.` symbol.
+//
+// All keys in the map will be prefixed with path.
+func DocumentFromAttributesWithPath(path string, am pdata.AttributeMap) Document {
+	if am.Len() == 0 {
+		return Document{}
+	}
+
+	fields := make([]field, 0, am.Len())
+	fields = appendAttributeFields(fields, path, am)
+	return Document{fields}
+}
+
+// AddTimestamp adds a raw timestamp value to the Document.
+func (doc *Document) AddTimestamp(key string, ts pdata.Timestamp) {
+	doc.Add(key, TimestampValue(ts.AsTime()))
+}
+
+// Add adds a converted value to the document.
+func (doc *Document) Add(key string, v Value) {
+	doc.fields = append(doc.fields, field{key: key, value: v})
+}
+
+// AddString adds a string to the document.
+func (doc *Document) AddString(key string, v string) {
+	if v != "" {
+		doc.Add(key, StringValue(v))
+	}
+}
+
+// AddID adds the hex presentation of an id value to the document. If the ID
+// is empty, no value will be added.
+func (doc *Document) AddID(key string, id idValue) {
+	if !id.IsEmpty() {
+		doc.AddString(key, id.HexString())
+	}
+}
+
+// AddInt adds an integer value to the document.
+func (doc *Document) AddInt(key string, value int64) {
+	doc.Add(key, IntValue(value))
+}
+
+// AddAttributes expands and flattens all key-value pairs from the input attribute map into
+// the document.
+func (doc *Document) AddAttributes(key string, attributes pdata.AttributeMap) {
+	doc.fields = appendAttributeFields(doc.fields, key, attributes)
+}
+
+// AddAttribute converts and adds a AttributeValue to the document. If the attribute represents a map,
+// the fields will be flattened.
+func (doc *Document) AddAttribute(key string, attribute pdata.AttributeValue) {
+	switch attribute.Type() {
+	case pdata.AttributeValueNULL:
+		// do not add 'null'
+	case pdata.AttributeValueMAP:
+		doc.AddAttributes(key, attribute.MapVal())
+	default:
+		doc.Add(key, ValueFromAttribute(attribute))
+	}
+}
+
+// Sort sorts all fields in the document by key name.
+func (doc *Document) Sort() {
+	sort.SliceStable(doc.fields, func(i, j int) bool {
+		return doc.fields[i].key < doc.fields[j].key
+	})
+
+	for i := range doc.fields {
+		fld := &doc.fields[i]
+		fld.value.Sort()
+	}
+}
+
+// Dedup removes fields from the document, that have duplicate keys.
+// The filtering only keeps the last value for a key.
+//
+// NOTE: The document MUST be sorted before dedup can be applied.
+func (doc *Document) Dedup() {
+	// 1. rename fields if a primitive value is overwritten by an object.
+	//    For example the pair (path.x=1, path.x.a="test") becomes:
+	//    (path.x.value=1, path.x.a="test").
+	//
+	//    This step removes potential conflicts when dedotting and serializing fields.
+	for i := 0; i < len(doc.fields)-1; i++ {
+		if len(doc.fields[i].key) < len(doc.fields[i+1].key) &&
+			strings.HasPrefix(doc.fields[i+1].key, doc.fields[i].key) {
+			doc.fields[i].key = doc.fields[i].key + ".value"
```

Comment:
Related to previous comment, this technically breaks the sorting invariant (puts "a.value" before "a.a") -- it looks like that's irrelevant, but it would be nice if a comment confirmed it :-)

Reply:
You are right. This can lead to a parse failure in ES and must be handled if we don't want to drop the event.

Reply:
Fix the sorting + fixed unit test